include/kernels/gauge_force.cuh

```diff
@@ -78,116 +27,72 @@ namespace quda {
 
-      int path0 = path[0];
-      int lnkdir = isForwards(path0) ? path0 : flipDir(path0);
+        // the gauge path starts pre-shifted, so we need to do the shift + update the parity
+        dx[dir]++;
+        int nbr_oddbit = (parity ^ 1);
 
-      if (isForwards(path0)) {
-        linkB = arg.u(lnkdir, linkIndexShift(x,dx,arg.E), nbr_oddbit);
-        linkA = linkB;
-        dx[lnkdir]++; // now have to update location
-	nbr_oddbit = nbr_oddbit^1;
-      } else {
-        dx[lnkdir]--; // if we are going backwards the link is on the adjacent site
-        nbr_oddbit = nbr_oddbit^1;
-	linkB = arg.u(lnkdir, linkIndexShift(x,dx,arg.E), nbr_oddbit);
-        linkA = conj(linkB);
-      }
+        // compute the path
+        link_prod = computeGaugePath(arg, x, nbr_oddbit, path, arg.p.length[i], dx);
 
-      for (int j=1; j<arg.p.length[i]; j++) {
-
-        int pathj = path[j];
-        int lnkdir = isForwards(pathj) ? pathj : flipDir(pathj);
-
-        if (isForwards(pathj)) {
-          linkB = arg.u(lnkdir, linkIndexShift(x,dx,arg.E), nbr_oddbit);
-          linkA = linkA * linkB;
-          dx[lnkdir]++; // now have to update to new location
-          nbr_oddbit = nbr_oddbit^1;
-        } else {
-          dx[lnkdir]--; // if we are going backwards the link is on the adjacent site
-	  nbr_oddbit = nbr_oddbit^1;
-          linkB = arg.u(lnkdir, linkIndexShift(x,dx,arg.E), nbr_oddbit);
-          linkA = linkA * conj(linkB);
-        }
-      } //j
-      staple = staple + coeff*linkA;
-    } //i
-
-    // multiply by U(x)
-    linkA = arg.u(dir, linkIndex(x,arg.E), parity);
-    linkA = linkA * staple;
-
-    // update mom(x)
-    Link mom = arg.mom(dir, idx, parity);
-    if(arg.compute_force) {
-      mom = mom - arg.epsilon * linkA;
-      makeAntiHerm(mom);
-    }
-    else {
-      mom = mom + arg.epsilon * linkA;
-    }
-    arg.mom(dir, idx, parity) = mom;
-  }
+        accum = accum + coeff * link_prod;
+      } //i
 
-  template <typename Arg> struct GaugeForce
-  {
-    const Arg &arg;
-    constexpr GaugeForce(const Arg &arg) : arg(arg) {}
-    static constexpr const char *filename() { return KERNEL_FILE; }    
+      // multiply by U(x)
+      link_prod = arg.u(dir, linkIndex(x,arg.E), parity);
+      link_prod = link_prod * accum;
 
-    __device__ __host__ void operator()(int x_cb, int parity, int dir)
-    {
-      switch(dir) {
-      case 0: GaugeForceKernel<Arg,0>(arg, x_cb, parity); break;
-      case 1: GaugeForceKernel<Arg,1>(arg, x_cb, parity); break;
-      case 2: GaugeForceKernel<Arg,2>(arg, x_cb, parity); break;
-      case 3: GaugeForceKernel<Arg,3>(arg, x_cb, parity); break;
+      // update mom(x)
+      Link mom = arg.mom(dir, x_cb, parity);
+      if (arg.compute_force) {
+        mom = mom - arg.epsilon * link_prod;
+        makeAntiHerm(mom);
+      } else {
+        mom = mom + arg.epsilon * link_prod;
       }
+      arg.mom(dir, x_cb, parity) = mom;
```

Comment:
Nice cleanup in this file.  Good to confirm that there's not performance difference.

Reply:
I believe I checked and it was sub-percentile, if any.